libs/wingc/src/type_check.rs

```diff
@@ -1803,16 +1793,21 @@ impl<'a> TypeChecker<'a> {
 		}
 	}
 
-	fn type_check_statement(&mut self, stmt: &Stmt, env: &mut SymbolEnv, context: &TypeCheckerContext) {
+	fn type_check_statement(&mut self, stmt: &Stmt, env: &mut SymbolEnv) {
+		// Set the current statement index for symbol lookup checks. We can safely assume we're
+		// not overwriting the current statement index because `type_check_statement` is never
+		// recursively called (we use a breadth-first traversal of the AST statements).
+		self.statement_idx = stmt.idx;
```

Comment:
hmm, maybe I still don't totally grasp how the breadth-first traversal works. How does a case like this get handled?

```
let a = 1;
let do_stuff = () => {
  let x = 2;
  let adder = (): num => {
    return a + b + x + y; // HERE
  };
  let y = 3;
  adder();
}
let b = 4;
```

I guess initially I thought this would require calling `type_check_statement` within another `type_check_statement` call, but is this comment saying that is not what happens?

(FWIW it looks like this example produces the correct errors in our current version of Wing so that is good news 👍)

Reply:
1. Traverse statements in top level and create this env

| symb | idx |
| --- | --- |
| a | 0 |
| do_stuff | 1 |
| b | 2 |

during traversal we store the inner scope of `do_stuff` for future work

2. Traverse the `do_stuff` scope and create this env with env.staement_idx=1 because it was created in the top level scope's statement idx 1

| symb | idx |
| --- | --- |
| x | 0 |
| addr | 1 |
| y | 2 |
| addr | 3 |

during traversal we store the inner scope of `addr` for future work

3.  Traverse the `addr` scope and create this env eith env.satemt_idx = 1 (becase it was created in `do_stuff`'s second statement)

| symb | idx |
| --- | --- |

(^nothing defined here)

When we handle `a + b + x + y` we know that we're in index 0 of the `adder` scope. And we do the following lookups:
`a`: it's not in our scope, go up, now we're in `do_stuff` env so our index is set to 1 (`addr` env's index), we still can't find `a`, go up, now we're the top env so our index is set to 1 (`do_stuff` env's index), we find `a` with index 0, its index is less than 1 (our index) so the lookup succeeds.

`b`: it's not in our scope, go up, now we're in `do_stuff` env so our index is set to 1 (`addr` env's index), we still can't find `b`, go up, now we're the top env so our index is set to 1 (`do_stuff` env's index), we find `b` with index 2, its index is more than 1 (our index) so the lookup fails.

`x`: it's not in our scope, go up, now we're in `do_stuff` env so our index is set to 1 (`addr` env's index), we find `x` with index 0, its index is less than 1 (our index) so the lookup succeeds.

`y`: it's not in our scope, go up, now we're in `do_stuff` env so our index is set to 1 (`addr` env's index), we find `y` with index 2, its index more than 1 (our index) so the lookup fails.

Hope that was followable :)

Reply:
This explanation is super helpful!! Yeah I think I was a little confused on how the statement indices are stored and how the lookups are performed, but now it makes sense to me. Thanks for writing this up 😄

Reply:
Amazing, thanks for the detailed explanation